datafusion/benches/physical_plan.rs

```diff
@@ -0,0 +1,176 @@
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#[macro_use]
+extern crate criterion;
+use criterion::{BatchSize, Criterion};
+extern crate arrow;
+extern crate datafusion;
+
+use std::{iter::FromIterator, sync::Arc};
+
+use arrow::{
+    array::{ArrayRef, Int64Array, StringArray},
+    record_batch::RecordBatch,
+};
+use tokio::runtime::Runtime;
+
+use datafusion::physical_plan::{
+    collect,
+    expressions::{col, PhysicalSortExpr},
+    memory::MemoryExec,
+    sort_preserving_merge::SortPreservingMergeExec,
+};
+
+// Initialise the operator using the provided record batches and the sort key
+// as inputs. All record batches must have the same schema.
+fn sort_preserving_merge_operator(batches: Vec<RecordBatch>, sort: &[&str]) {
+    let schema = batches[0].schema();
+
+    let sort = sort
+        .iter()
+        .map(|name| PhysicalSortExpr {
+            expr: col(name, &schema).unwrap(),
+            options: Default::default(),
+        })
+        .collect::<Vec<_>>();
+
+    let exec = MemoryExec::try_new(
+        &batches.into_iter().map(|rb| vec![rb]).collect::<Vec<_>>(),
+        schema,
+        None,
+    )
+    .unwrap();
+    let merge = Arc::new(SortPreservingMergeExec::new(sort, Arc::new(exec), 10000));
+
+    let rt = Runtime::new().unwrap();
+    rt.block_on(collect(merge)).unwrap();
+}
+
+// Produces `n` record batches of row size `m`. Each record batch will have
+// identical contents except for if the `batch_offset` is set. In that case the
+// values for column "d" in each subsequent record batch will be offset in
+// value.
+//
+// The `rows_per_key` value controls how many rows are generated per "key",
+// which is defined as columns a, b and c.
+fn batches(
+    n: usize,
+    m: usize,
+    rows_per_sort_key: usize,
+    batch_offset: usize,
+) -> Vec<RecordBatch> {
+    let mut rbs = Vec::with_capacity(n);
+    let mut curr_batch_offset = 0;
+
+    for _ in 0..n {
+        let mut col_a = Vec::with_capacity(m);
+        let mut col_b = Vec::with_capacity(m);
+        let mut col_c = Vec::with_capacity(m);
+        let mut col_d = Vec::with_capacity(m);
+
+        let mut j = 0;
+        let mut current_rows_per_sort_key = 0;
+
+        for i in 0..m {
+            if current_rows_per_sort_key == rows_per_sort_key {
+                current_rows_per_sort_key = 0;
+                j = i;
+            }
+
+            col_a.push(Some(format!("a-{:?}", j)));
+            col_b.push(Some(format!("b-{:?}", j)));
+            col_c.push(Some(format!("c-{:?}", j)));
+            col_d.push(Some((i + curr_batch_offset) as i64));
+
+            current_rows_per_sort_key += 1;
+        }
+
+        col_a.sort();
```

Comment:
This data is only sorted on `a` correct? In other words, the `SortPreservingMerge` operator will only see sorted input data if the sort key is `["a"]`, but the benchmark is using `["a", "b", "c", "d"]` as the sort key. Or perhaps I am mis reading this

Reply:
I sorted the columns  independently because it was a bit simpler than invoking a lex sort on them and technically the rows are still sorted by `a, b, c, d`, it's just that the values for each column in the sort key change in unison.

For example to create two inputs where the sort keys don't overlap the test data might look like:

Input A:
```
+------+------+------+----+
| a    | b    | c    | d  |
+------+------+------+----+
| a-0  | b-0  | c-0  | 0  |
| a-0  | b-0  | c-0  | 1  |
| a-0  | b-0  | c-0  | 2  |
| a-0  | b-0  | c-0  | 3  |
| a-0  | b-0  | c-0  | 4  |
| a-0  | b-0  | c-0  | 5  |
| a-0  | b-0  | c-0  | 6  |
| a-0  | b-0  | c-0  | 7  |
| a-0  | b-0  | c-0  | 8  |
| a-0  | b-0  | c-0  | 9  |
| a-10 | b-10 | c-10 | 10 |
| a-10 | b-10 | c-10 | 11 |
| a-10 | b-10 | c-10 | 12 |
| a-10 | b-10 | c-10 | 13 |
| a-10 | b-10 | c-10 | 14 |
| a-10 | b-10 | c-10 | 15 |
| a-10 | b-10 | c-10 | 16 |
| a-10 | b-10 | c-10 | 17 |
| a-10 | b-10 | c-10 | 18 |
| a-10 | b-10 | c-10 | 19 |
```

Then using a `batch_offset = 10` the second input would be generated as:

```
+------+------+------+----+
| a    | b    | c    | d  |
+------+------+------+----+
| a-0   | b-0   | c-0   | 5    |
| a-0   | b-0   | c-0   | 6    |
| a-0   | b-0   | c-0   | 7    |
| a-0   | b-0   | c-0   | 8    |
| a-0   | b-0   | c-0   | 9    |
| a-0   | b-0   | c-0   | 10   |
| a-0   | b-0   | c-0   | 11   |
| a-0   | b-0   | c-0   | 12   |
| a-0   | b-0   | c-0   | 13   |
| a-0   | b-0   | c-0   | 14   |
| a-10  | b-10  | c-10  | 15   |
| a-10  | b-10  | c-10  | 16   |
| a-10  | b-10  | c-10  | 17   |
| a-10  | b-10  | c-10  | 18   |
| a-10  | b-10  | c-10  | 19   |
| a-10  | b-10  | c-10  | 20   |
| a-10  | b-10  | c-10  | 21   |
| a-10  | b-10  | c-10  | 22   |
| a-10  | b-10  | c-10  | 23   |
| a-10  | b-10  | c-10  | 24   |
```

In this case the operator would take the first five rows from input A then have to alternate taking the next five rows from input A and input B (for rows where `d = [5, 9]`. Then finally take all the rows from input B.

Then it would move back to input A for the row containing `| a-10 | b-10 | c-10 | 10 |` and do the same thing again.

The idea here is that in `main` right now merging the above data would involve (I think) 40 calls to `extend()` but this PR should make it so it only involves `2 * (1 + 10 + 1) = 24` calls.